### PR TITLE
Dht cas merge

### PR DESCRIPTION
**Pull request by ianopolous**
This PR merges the old DHT interface into the ContentAddressedStorage interface. In the process it matches our http protocol with IPFS's. It also moves the btree from a server side implementation to a client side one. And introduces caching for core node IPNS lookups. These are breaking changes for the data formats. They will also break again once I move to the ipfs dag api, which is cleaner than the object api which is now deprecated.

### CODE REVIEW

File: src/peergos/server/net/DHTHandler.java [hunk 1]
```diff
@@ -1,120 +1,174 @@
...
+    }
+
+    private Map<String, List<String>> parseQuery(String query) {
+        if (query.startsWith("?"))
+            query = query.substring(1);
+        String[] parts = query.split("&");
+        Map<String, List<String>> res = new HashMap<>();
+        for (String part: parts) {
+            int sep = part.indexOf("=");
+            String key = part.substring(0, sep);
```
Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: key and  value should be url-decoded.

Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: Also this  could be  abstracted into a tested utility lib.

File: src/peergos/server/net/DHTHandler.java [hunk 2]
```diff
@@ -1,120 +1,174 @@
...
-                    SortedMap<String, Multihash> namedLinks = new TreeMap<>();
-                    for (int i = 0; i < hashes.size(); i++)
-                        namedLinks.put(Integer.toString(i), hashes.get(i));
-                    MerkleNode obj = new MerkleNode(value, namedLinks);
-                    byte[] put = dht.put(obj).toBytes();
-                    new PutSuccess(httpExchange).accept(Optional.of(put));
-                } else {
+            switch (path) {
+                case "object/new":{
+                    UserPublicKey writer = UserPublicKey.fromString(last.apply("writer"));
```
Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: This switch makes the function a bit large. Perhaps each case could be  be  handled by a separate method?

File: src/peergos/server/net/DHTHandler.java [hunk 3]
```diff
@@ -1,120 +1,174 @@
...
+                    dht._new(writer).thenAccept(newHash -> {
+                        Map res = new HashMap();
```
Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: Could this be typed?

File: src/peergos/server/net/DHTHandler.java [hunk 4]
```diff
@@ -1,120 +1,174 @@
...
-                    byte[] value = Serialize.deserializeByteArray(din, ContentAddressedStorage.MAX_OBJECT_LENGTH);
-                    byte[] owner = Serialize.deserializeByteArray(din, UserPublicKey.MAX_SIZE);
-                    // TODO check we care about owner
-                    List<Multihash> hashes = new ArrayList<>();
-                    int nlinks = din.readInt();
-                    for (int i = 0; i < nlinks; i++)
-                        hashes.add(new Multihash(Serialize.deserializeByteArray(din, 1024)));
+            if (! path.startsWith(apiPrefix))
+                throw new IllegalStateException("Unsupported api version, required: " + apiPrefix);
+            path = path.substring(apiPrefix.length());
```
Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: It would be nice  to abstract this  logic into a re-usable super-class, something like a VersionedHTTPHandler.

File: src/peergos/server/net/DHTHandler.java [hunk 5]
```diff
@@ -1,120 +1,174 @@
...
+                    break;
+                }
+                case "object/patch/add-link":{
+                    UserPublicKey writer = UserPublicKey.fromString(last.apply("writer"));
+                    Multihash hash = Multihash.fromBase58(args.get(0));
+                    String label = args.get(1);
+                    Multihash targetHash = Multihash.fromBase58(args.get(2));
+                    dht.addLink(writer, hash, label, targetHash)
+                            .thenAccept(resultHash -> {
+                                Map res = new HashMap();
```
Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: Could this be  typed?

File: src/peergos/server/net/DHTHandler.java [hunk 6]
```diff
@@ -1,120 +1,174 @@
...
+                    break;
+                }
+                default: {
                     httpExchange.sendResponseHeaders(404, 0);
                 }
             }
         } catch (Exception e) {
             e.printStackTrace();
-            throw e;
+            throw new RuntimeException(e);
```
Comment (cboddy) [inline src/peergos/server/net/DHTHandler.java]: It would be nice to  send  back an internal-server-error to the client here.

File: src/peergos/server/net/MultipartReceiver.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+public class MultipartReceiver {
+
+    public static byte[] extractFile(InputStream in, String boundary) {
+        try {
+            String first = readLine(in);
+            if (!first.substring(2).equals(boundary))
+                throw new IllegalStateException("Incorrect boundary! " + boundary + " != " + first.substring(2));
+            String filenameLine = readLine(in);
+            String contentType = readLine(in);
```
Comment (cboddy) [inline src/peergos/server/net/MultipartReceiver.java]: This looks  super useful  and re-usable. Could we add a test for the basic functionality and error conditions?

File: src/peergos/server/net/MultipartReceiver.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                file.write(buffer, 0, r);
+            byte[] bytes = file.toByteArray();
+            // 8 is two \r\n and two -- at end of stream
+            return Arrays.copyOfRange(bytes, 0, bytes.length - (8 + boundary.length()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static String readLine(InputStream in) throws IOException {
```
Comment (cboddy) [inline src/peergos/server/net/MultipartReceiver.java]: Perhaps this  function should  take a  max-length parameter like the  Serialize.deserializeString method to stop reading an  infinitely long stream?

File: src/peergos/server/storage/CachingStorage.java [hunk 9]
```diff
@@ -1,43 +1,68 @@
...
         this.target = target;
-        this.cache = new LRUCache(cacheSize);
+        this.cache = new LRUCache<>(cacheSize);
         this.maxValueSize = maxValueSize;
     }
 
     @Override
-    public Multihash put(MerkleNode object) {
-        return target.put(object);
+    public CompletableFuture<Multihash> _new(UserPublicKey writer) {
```
Comment (cboddy) [inline src/peergos/server/storage/CachingStorage.java]: Could we  call this something more descriptive like `createEmpty`?

File: src/peergos/server/storage/IpfsDHT.java [hunk 10]
```diff
@@ -27,54 +30,94 @@ public IpfsDHT() {
     }
 
     @Override
-    public byte[] get(Multihash key) {
+    public CompletableFuture<Multihash> _new(UserPublicKey writer) {
```
Comment (cboddy) [inline src/peergos/server/storage/IpfsDHT.java]: Again, could we call this something else?

File: src/peergos/server/storage/RAMStorage.java [hunk 11]
```diff
@@ -1,27 +1,64 @@
...
 import java.util.*;
+import java.util.concurrent.*;
 
 public class RAMStorage implements ContentAddressedStorage {
     private Map<Multihash, byte[]> storage = new HashMap<>();
 
+    private final Set<Multihash> pinnedRoots = new HashSet<>();
+
+    @Override
+    public CompletableFuture<Multihash> _new(UserPublicKey writer) {
```
Comment (cboddy) [inline src/peergos/server/storage/RAMStorage.java]: Not a big fan of this name.

File: src/peergos/server/tests/MerkleNodeSerialization.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import peergos.shared.merklebtree.*;
+
+import java.io.*;
+import java.util.*;
+
+public class MerkleNodeSerialization {
+
+    @Test
+    public void inverses() throws IOException {
+        MerkleNode original = new MerkleNode(new byte[0], Collections.emptyList());
```
Comment (cboddy) [inline src/peergos/server/tests/MerkleNodeSerialization.java]: Could we add a test for a non empty MerkleNode?

File: src/peergos/server/storage/CachingStorage.java (hunk 9 above)
Comment (ianopolous) [inline src/peergos/server/storage/CachingStorage.java]: this is just matching the ipfs name which is new

File: src/peergos/server/storage/IpfsDHT.java (hunk 10 above)
Comment (ianopolous) [inline src/peergos/server/storage/IpfsDHT.java]: same as ipfs

File: src/peergos/server/storage/RAMStorage.java (hunk 11 above)
Comment (ianopolous) [inline src/peergos/server/storage/RAMStorage.java]: just matching ipfs

File: src/peergos/server/tests/MerkleNodeSerialization.java (hunk 12 above)
Comment (ianopolous) [inline src/peergos/server/tests/MerkleNodeSerialization.java]: done

File: src/peergos/server/net/DHTHandler.java (hunk 4 above)
Comment (ianopolous) [inline src/peergos/server/net/DHTHandler.java]: I looked into this, and because of the Java api requiring a HttpExchange this wouldn't actually encapsulate anything, The httpExchange contains the full path anyway. In essence to do this would require writing our own web framework, so basically jet all over again. So I think it's simpler how it is.

File: src/peergos/server/net/DHTHandler.java (hunk 1 above)
Comment (ianopolous) [inline src/peergos/server/net/DHTHandler.java]: They already are url decoded. URI.getQuery() does this

File: src/peergos/server/net/MultipartReceiver.java (hunk 7 above)
Comment (ianopolous) [inline src/peergos/server/net/MultipartReceiver.java]: this is already tested by all the user tests

File: src/peergos/server/net/DHTHandler.java (hunk 6 above)
Comment (ianopolous) [inline src/peergos/server/net/DHTHandler.java]: done

File: src/peergos/server/net/DHTHandler.java (hunk 5 above)
Comment (ianopolous) [inline src/peergos/server/net/DHTHandler.java]: it is json and it will change when we move to the dag api

File: src/peergos/server/tests/MerkleNodeSerialization.java (hunk 12 above)
Comment (ianopolous) [inline src/peergos/server/tests/MerkleNodeSerialization.java]: done

File: src/peergos/server/storage/RAMStorage.java (hunk 11 above)
Comment (ianopolous) [inline src/peergos/server/storage/RAMStorage.java]: ok. done :-)
